### PR TITLE
Implement integer to long/ulong casts for x86

### LINKED ISSUE #4664: RyuJIT x86: NYI: "Lowering of signed cast TYP_INT->TYP_LONG"
JIT\Methodical\MDArray\DataTypes\long_cs_d\long_cs_d.cmd
JIT\Methodical\MDArray\DataTypes\long_cs_do\long_cs_do.cmd
JIT\Methodical\MDArray\DataTypes\long_cs_r\long_cs_r.cmd
JIT\Methodical\MDArray\DataTypes\long_cs_ro\long_cs_ro.cmd

sample test output:

```
ERROR:
result is: -1
2darr[0, 1] is: 4294967295
and they are NOT equal !

ERROR:
result is: -1
cl1.long2darr[0, 1] is: 4294967295
and they are NOT equal !

ERROR:
result is: -1
long3darr[1,0,1] is: 4294967295
and they are NOT equal !

ERROR:
result is: -1
cl1.long3darr[1,0,1] is: 4294967295
and they are NOT equal !

ERROR:
Decimal_result is: -1
2darr[0, 1] is: 4294967295
and they are NOT equal !

ERROR:
Decimal_result is: -1
cl1.long2darr[0, 1] is: 4294967295
and they are NOT equal !

ERROR:
Decimal_result is: -1
long3darr[1,0,1] is: 4294967295
and they are NOT equal !

ERROR:
Decimal_result is: -1
cl1.long3darr[1,0,1] is: 4294967295
and they are NOT equal !

ERROR:
Double_result is: -1
2darr[0, 1] is: 4294967295
and they are NOT equal !

ERROR:
Double_result is: -1
cl1.long2darr[0, 1] is: 4294967295
and they are NOT equal !

ERROR:
Double_result is: -1
long3darr[1,0,1] is: 4294967295
and they are NOT equal !

ERROR:
Double_result is: -1
cl1.long3darr[1,0,1] is: 4294967295
and they are NOT equal !

ERROR:
Single_result is: -1
2darr[0, 1] is: 4294967295
and they are NOT equal !

ERROR:
Single_result is: -1
cl1.long2darr[0, 1] is: 4294967295
and they are NOT equal !

ERROR:
Single_result is: -1
long3darr[1,0,1] is: 4294967295
and they are NOT equal !

ERROR:
Single_result is: -1
cl1.long3darr[1,0,1] is: 4294967295
and they are NOT equal !


Unhandled Exception: System.OverflowException: Value was either too large or too small for an Int32.
   at System.Convert.ThrowInt32OverflowException()
   at System.Convert.ToInt32(Int64 value)
   at longMDArrTest.Main()
Expected: 100
Actual: -532462766
```

### PR DESCRIPTION
**Pull request by mikedn**
Implement various integer to long/ulong casts for x86. Sign extension is performed using an arithmetic shift to avoid the need for the x86 specific `cdq`. Overflow checks are implemented by using int->uint casts for which codegen provides the necessary overflow support.

Fixes #4664, #4174, #4175

### CODE REVIEW

**Comment by mikedn** [comment]
@BruceForstall In addition to the sign extension issue I also added support for casts with overflow checks, initially I wasn't sure how well it will work but it seems easy enough.

File: src/jit/lower.cpp [hunk 1]
```diff
@@ -303,7 +303,7 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
 
             hiStore->gtOp.gtOp1 = hiRhs;
             hiStore->CopyCosts(tree);
-            hiStore->gtFlags |= GTF_VAR_DEF;
+            hiStore->gtFlags = tree->gtFlags;
```
Comment (mikedn) [inline src/jit/lower.cpp]: I'm not sure this is the proper fix. The issue is that the `hiStore` tree lacks the GTF_EXCEPT flag that overflow checks add and an assert fires. Simply copying all flags appears to work fine.

File: src/jit/lower.cpp [hunk 2]
```diff
@@ -338,24 +338,96 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
...
+                    // Change loResult and hiResult to null so the code after the switch doesn't 
+                    // attempt to insert hiResult after loResult. We only need it to insert newTree.
+                    //
+
+                    loResult = nullptr;
+                    hiResult = nullptr;
+                }
                 else
                 {
-                    NYI("Lowering of signed cast TYP_INT->TYP_LONG");
```
Comment (mikedn) [inline src/jit/lower.cpp]: I suspect that this shouldn't be hit so I didn't bother with it.

File: src/jit/lower.cpp [hunk 3]
```diff
@@ -338,24 +338,96 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
...
+                    NYI("Unimplemented long->long no-op cast");
                 }
-                break;
-            default:
+            }
+            else
+            {
                 NYI("Unimplemented type for Lowering of cast to TYP_LONG");
-                break;
```
Comment (mikedn) [inline src/jit/lower.cpp]: The main remaining types are float and double, such casts should be implemented by helper calls so it's likely that this won't be hit either.

**Comment by BruceForstall** [comment]
Looks like you have a few relevant test failures in the x86 ryujit suite.

**Comment by BruceForstall** [comment]
I'll look (might be tomorrow).

Also, @adiaaida PTAL
/cc @dotnet/jit-contrib

File: src/jit/lower.cpp (hunk 1 above)
Comment (CarolEidt) [inline src/jit/lower.cpp]: I think that the right fix is `hiStore->gtFlags |= (tree->gtFlags & (GTF_ALL_EFFECT|GTF_LIVENESS_MASK));`

File: src/jit/lower.cpp [hunk 4]
```diff
@@ -338,24 +338,96 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
...
+                    loResult->AsCast()->gtCastType = TYP_UINT;
+                    loResult->gtType = TYP_INT;
+
+                    hiResult = comp->gtNewIconNode(0);
+                }
+                else
                 {
+                    comp->fgInsertEmbeddedFormTemp(&tree->gtOp.gtOp1);
+
                     loResult = tree->gtGetOp1();
```
Comment (adiaaida) [inline src/jit/lower.cpp]: Why aren't we using srcOp here?

Comment (mikedn) [inline src/jit/lower.cpp]: @adiaaida `fgInsertEmbeddedFormTemp` replaces the operand, probably I should add a `srcOp = tree->gtGetOp1()` right after `fgInsertEmbeddedFormTemp` for clarity. Or just get rid of `srcOp` completely.

File: src/jit/lower.cpp [hunk 5]
```diff
@@ -338,24 +338,96 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
...
+            var_types dstType = tree->AsCast()->CastToType();
+
+            if (tree->gtFlags & GTF_UNSIGNED)
             {
-            case TYP_INT:
-                if (tree->gtFlags & GTF_UNSIGNED)
+                srcType = genUnsignedType(srcType);
+            }
+
+            if (genActualTypeIsIntOrI(srcType) || varTypeIsGC(srcType))
```
Comment (CarolEidt) [inline src/jit/lower.cpp]: I think this could/should be simply `varTypeIsIntegralOrI(srcType)`. I just noticed that both `varTypeIsIntOrI()` and `genActualTypeIsIntOrI()` don't actually include types with the `VTF_I` classification, which seems at a minimum quite confusing.

Comment (CarolEidt) [inline src/jit/lower.cpp]: OK, I now see that this was actually intended to exclude long types, so I take it back. A comment might be good to clarify that point (i.e. that we're looking for source types that fit in a single 32-bit register).

Comment (mikedn) [inline src/jit/lower.cpp]: I think I could change those `if`s as follows:

```
if (varTypeIsLong(srcType))
{
    ...
}
else if (varTypeIsIntegralOrI(srcType))
{
    ...
}
else
{
    NYI
}
```

I'll take a look once I figure out what's up with those failing tests.

File: src/jit/lower.cpp [hunk 6]
```diff
@@ -3953,6 +4025,16 @@ void Lowering::DoPhase()
...
+        // Decompose TYP_LONG nodes
+        for (stmt = block->bbTreeList; stmt != nullptr; stmt = stmt->gtNext)
+        {
+            comp->compCurStmt = stmt;
+            comp->fgWalkTreePost(&stmt->gtStmt.gtStmtExpr, &Lowering::DecompNodeHelper, this, true);
+            stmt = comp->compCurStmt;
+        }
+#endif
+
         /* Walk the statement trees in this basic block */
```
Comment (CarolEidt) [inline src/jit/lower.cpp]: Is the reason for pulling this out into a separate loop because the decomposition might turn a single statement into two?

Comment (mikedn) [inline src/jit/lower.cpp]: Yes, embedded statements created by `fgInsertEmbeddedFormTemp` aren't lowered without this change.

But as mentioned in another issue I'm wondering if decomposition shouldn't happen after lowering. That may simplify GT_IND handling because after lowering we have GT_LEA nodes.

Comment (CarolEidt) [inline src/jit/lower.cpp]: Actually, I think a better solution is to move the addressing mode realization earlier in the phase order. We already have CSE determining whether an addressing mode is likely to be created, before it decides whether to create a CSE. This is something we've discussed, but it hasn't bubbled up to the top of the list.

**Comment by mikedn** [comment]
The assert that fires in test JIT\Regression\JitBlue\DevDiv_200492\DevDiv_200492\DevDiv_200492.cmd is interesting. The generated code looks like:
```
G_M30135_IG01:

G_M30135_IG02:
       33C0         xor      eax, eax
       50           push     eax                                   ; push loResult
       83FA00       cmp      edx, 0                                ; overflow check for hiResult
       7C0A         jl       SHORT G_M30135_IG04                   ; exit bb
       8BC2         mov      eax, edx
       50           push     eax                                   ; push hiResult
       FF15FC7BF902 call     [TestUlongAttribute:Store(long):this]

G_M30135_IG03:
       C3           ret      

G_M30135_IG04:
       E87199AA50   call     CORINFO_HELP_OVERFLOW
       CC           int3     
```
I doubt that it's valid to throw when `esp` doesn't point to the current frame and that's probably what the assert is trying to tell. Hoisting the overflow check out of the call is unlikely to work in general case as it will affect evaluation order so I guess that such method require a `ebp` frame.

**Comment by mikedn** [comment]
Test JIT\Methodical\fp\exgen\10w5d_cs_do\10w5d_cs_do.cmd is a nightmare. It fails left and right depending on options such as JitInline and JitMinOpts. For example, if I use JitMinOpts=1 I get some code like the following:
```
       BA02000000   mov      edx, 2
       E8F993E8FC   call     System.Single[,]:Get(int,int):float:this
       D95DF4       fstp     dword ptr [ebp-0CH]
       F30F1045F4   movss    xmm0, dword ptr [ebp-0CH]
       F30F5CC8     subss    xmm1, xmm0          ; xmm1 wasn't defined anywhere
       F30F114DF4   movss    dword ptr [ebp-0CH], xmm1
```
This might have something to do with float/double returns but I don't know what this has to do with my cast changes. The method where this happens doesn't even use long operations.

**Comment by BruceForstall** [comment]
Well, RyuJIT x86 is pretty raw...

Try using `COMPLUS_AltJit=foo` to only compile foo with RyuJIT/x86. Use `COMPLUS_JitFunctionTrace=1` and `COMPLUS_JitFuncInfoLogFile=log.txt` to see what functions hit NYI and which don't. Sometimes it's confusing because RyuJIT/x86 with optimization will inline something and succeed, but without optimization will hit an NYI.

**Comment by mikedn** [comment]
For now I created #5047 to deal with the bad float return issue. I'll have to double check but the nightmare test seems to pass if that issue is fixed.

**Comment by mikedn** [comment]
Found another issue with long operations while investigating JIT\Methodical\fp\exgen\10w5d_cs_do\10w5d_cs_do.cmd. Sometimes lowering replaces an `add` with a `lea`:
```
       81EA3BD6A42A sub      edx, 0x2AA4D63B
       B90087947F   mov      ecx, 0x7F948700
       1BC1         sbb      eax, ecx
       8D0C16       lea      ecx, [esi+edx] 
       8BD7         mov      edx, edi
       13D0         adc      edx, eax
       8BC1         mov      eax, ecx
```

**Comment by mikedn** [comment]
Tests pass if workarounds are added for the 3 unrelated bugs I've found. However, this implementation is still flawed, the following test asserts:
```
class Program
{
    int i;

    [MethodImpl(MethodImplOptions.NoInlining)]
    ulong Test(int h)
    {
        int x = h + 4;
        ulong f = checked((ulong)unchecked(x - i));
        if (i > h)
            return 0;
        return f;
    }

    static void Main()
    {
        new Program().Test(42);
    }
}
```
Test is adapted from the test I made for the div/mod assert so the problem is likely similar - using `fgInsertEmbeddedFormTemp` while lowering/decomposing.

**Comment by mikedn** [comment]
Turns out that the above issue is caused by a bug in `GT_STORE_LCL_VAR` decomposition. It tries to insert a new statement after the current one but the current statement has an embedded statement. The new statement ends up being inserted between the current statement and its embedded statement.

How do we move forward with all this stuff? Attempting to add support for long casts uncovered 4 unrelated issues. Probably the same thing will happen if we try to add support for other long operations.

File: src/jit/lsra.cpp [hunk 7]
```diff
@@ -4115,6 +4115,8 @@ LinearScan::setFrameType()
         }
     }
 
+    frameType = FT_EBP_FRAME;
+
```
Comment (CarolEidt) [inline src/jit/lsra.cpp]: Why is this unconditionally setting the frame type to EBP?

**Comment by CarolEidt** [comment]
@mikedn - Given where we are with supporting longs on 32-bit, I think that the only near-term way forward is to deal with each of these issues in turn; I wish I could give a more definitive answer, but I think some of what we are doing is postulating a way forward, and then discovering the issues.
For the `GT_STORE_LCL_VAR` issue, I suspect the answer is possibly messy, but straightforward - we need to ensure that the embedded case is correctly handled. For the lea issue, I think the right answer is to have a flag something like `GTF_FLAGS` that indicates that these nodes have a flag dependency. (Ouch - too confusing talking about two kinds of flags.) Setting that would also allow us to avoid generating an lea in that case.

File: src/jit/lsra.cpp (hunk 7 above)
Comment (mikedn) [inline src/jit/lsra.cpp]: Temporary workaround to get JIT\Regression\JitBlue\DevDiv_200492\DevDiv_200492\DevDiv_200492.cmd working, see explanation a few posts above. The correct solution is to detect when an EBP frame is required for x86. Probably I should create an issue for this.

Obviously, those workarounds need to be removed if this is to be merged.

**Comment by mikedn** [comment]
> Given where we are with supporting longs on 32-bit, I think that the only near-term way forward is to deal with each of these issues in turn; I wish I could give a more definitive answer, but I think some of what we are doing is postulating a way forward, and then discovering the issues.

I suppose separate issues should be created and addressed individually rather than attempting to fix everything in a single PR?

> For the  GT_STORE_LCL_VAR  issue, I suspect the answer is possibly messy, but straightforward - we need to ensure that the embedded case is correctly handled.

Might be easier to always put the hiStore in an embedded statement before the loStore, even if this means that those stores don't follow the usual lo/hi order. Anyway I'm not sure how the current code works at all, it would seem that dragging hiRhs in a separate statement would cause a lot of problems if hiRhs is not a leaf.

> For the lea issue, I think the right answer is to have a flag something like  GTF_FLAGS  that indicates that these nodes have a flag dependency.

Or adding a `GT_ADD_LO`.

**Comment by mikedn** [comment]
> For the lea issue, I think the right answer is to have a flag something like GTF_FLAGS that indicates that these nodes have a flag dependency.

Or perhaps we can piggyback on the existing but presently unused `GTF_SET_FLAGS`. And perhaps later that one can also be used to eliminate the unnecessary 0 compare from code like `if (x + 42 == 0)`.

**Comment by CarolEidt** [comment]
I hadn't realized that `GTF_SET_FLAGS` existed. I think that would be a good choice.
> I suppose separate issues should be created and addressed individually rather than attempting to fix everything in a single PR?
That would be great. Without your changes, can you duplicate the `GT_STORE_LCL_VAR` issue (even if it doesn't result in an assert or bad codegen)?

**Comment by mikedn** [comment]
> Without your changes, can you duplicate the  GT_STORE_LCL_VAR  issue (even if it doesn't result in an assert or bad codegen)?

See #5696

**Comment by mikedn** [comment]
I'll change this to [WIP] until various other issues are fixed. This depends on #5698, #5047, #5699 and maybe a few others. And it looks like there's another problem with `GT_STORE_LCL_VAR` decomposition, different from the one that was fixed recently.

**Comment by BruceForstall** [comment]
@mikedn: I'm a little confused now about the overall status of this PR. Can you summarize?

**Comment by mikedn** [comment]
@BruceForstall Cast decomposition itself should be fine. I'm just trying to workaround a few other x86 issues (#6004 in particular) to get a clean test run.

File: src/jit/decomposelongs.cpp [hunk 8]
```diff
@@ -532,25 +523,92 @@ void DecomposeLongs::DecomposeCast(GenTree** ppTree, Compiler::fgWalkData* data)
...
+    if (varTypeIsLong(srcType))
     {
-    case TYP_INT:
-        if (tree->gtFlags & GTF_UNSIGNED)
+        if (tree->gtOverflow() && (varTypeIsUnsigned(srcType) != varTypeIsUnsigned(dstType)))
         {
-            loResult = tree->gtGetOp1();
-            hiResult = new (m_compiler, GT_CNS_INT) GenTreeIntCon(TYP_INT, 0);
-            m_compiler->fgSnipNode(curStmt, tree);
+            GenTree* srcOp = tree->gtGetOp1();
```
Comment (sejongoh) [inline src/jit/decomposelongs.cpp]: Are long -> int casts decomposed here too?

Comment (sejongoh) [inline src/jit/decomposelongs.cpp]: Oh.. I forgot that cast to int is returned early before entering DecomposeCast. This seems to be a little confusing.

Comment (mikedn) [inline src/jit/decomposelongs.cpp]: Yes, it's kind of confusing and I'm not 100% sure it happens all the time. I suppose it's done like this because otherwise we'd end up with a dangling hi tree.

File: src/jit/decomposelongs.cpp [hunk 9]
```diff
@@ -526,25 +526,92 @@ void DecomposeLongs::DecomposeCast(GenTree** ppTree, Compiler::fgWalkData* data)
...
+            // check provided by codegen.
+            //
+
+            loResult = loSrcOp;
+
+            hiResult                       = tree;
+            hiResult->gtType               = TYP_INT;
+            hiResult->AsCast()->gtCastType = TYP_UINT;
+            hiResult->gtFlags &= ~GTF_UNSIGNED;
+            hiResult->gtOp.gtOp1 = hiSrcOp;
```
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: Vertical alignment is ridiculous :-1:

**Comment by mikedn** [comment]
I did a merge quick merge after the LIR changes were committed and it looks like I messed up pretty badly as a large number of tests are failing. I'm on vacation for a week, I'll fix this when I get back.

File: src/jit/decomposelongs.cpp [hunk 10]
```diff
@@ -401,25 +404,40 @@ GenTree* DecomposeLongs::DecomposeStoreLclVar(LIR::Use& use)
...
     GenTree* loRhs   = rhs->gtGetOp1();
     GenTree* hiRhs   = rhs->gtGetOp2();
+
+    // Keep the original store's flags except side effects, the store
+    // itself should not have any. The lo and hi rhs parts could, at least
+    // in theory, have different side effects so we can't simply copy the
+    // original store's side effecs to both parts.
+    unsigned flags   = tree->gtFlags & ~GTF_ALL_EFFECT;
+    unsigned loFlags = flags | (loRhs->gtFlags & GTF_ALL_EFFECT);
+    unsigned hiFlags = flags | (hiRhs->gtFlags & GTF_ALL_EFFECT);
```
Comment (pgavlin) [inline src/jit/decomposelongs.cpp]: The store nodes should also have the `GTF_ASG` flag set. We should probably validate that all stores have this flag set in `CheckLIR`, but that change should be made separately.

**Comment by pgavlin** [comment]
I think this is very close: the changes at https://gist.github.com/pgavlin/77e2364408d6e63223b19319757c2170 take care of the cost-related asserts.

**Comment by mikedn** [comment]
@pgavlin Are the cost related changes still needed? If I understand correctly the costs are no longer needed since #6849 but I see that decomposition still attempts to copy costs here and there.

**Comment by pgavlin** [comment]
@mikedn: the cost changes should no longer be necessary, no. If you notice any parts of the backend that still attempt to maintain cost info, those parts should be fixed.

**Comment by pgavlin** [comment]
(indeed, attempting to maintain cost information may even cause asserts if the source node was added or modified by rationalize, which no longer sets or maintains cost info)

**Comment by mikedn** [comment]
The failure of `JIT\Performance\CodeQuality\BenchI\Pi\Pi\Pi.cmd` appears to be caused by a bug in long compare codegen. The generation of the compare is delayed until `GT_JTRUE` but that only works correctly if the jump follows the compare. In the case of PI benchmark LSRA appears to have inserted 2 reloads between the compare and the jump and those 2 reloads modify registers that the compare is supposed to be using.

**Comment by mikedn** [comment]
Well, this one is ready provided that we decide what to do about the failing test. It may sound odd but I'd fix the test because that long comparison is probably an accident. #7038 contains a much simpler repro that we can later add as a separate test.

File: src/jit/decomposelongs.cpp [hunk 11]
```diff
@@ -494,32 +494,100 @@ GenTree* DecomposeLongs::DecomposeCast(LIR::Use& use)
...
+        {
+            NYI("Unimplemented long->long no-op cast decomposition");
+        }
+    }
+    else if (varTypeIsIntegralOrI(srcType))
     {
-        case TYP_INT:
-            if (tree->gtFlags & GTF_UNSIGNED)
+        if (varTypeIsUnsigned(srcType))
+        {
```
Comment (pgavlin) [inline src/jit/decomposelongs.cpp]: Nit: might be nice to have a note in here that this is a conversion (u32 -> u64) that cannot overflow.

File: src/jit/decomposelongs.cpp [hunk 12]
```diff
@@ -494,32 +494,100 @@ GenTree* DecomposeLongs::DecomposeCast(LIR::Use& use)
...
 
-                hiResult = new (m_compiler, GT_CNS_INT) GenTreeIntCon(TYP_INT, 0);
                 Range().InsertAfter(loResult, hiResult);
             }
             else
             {
-                NYI("Lowering of signed cast TYP_INT->TYP_LONG");
-            }
-            break;
+                LIR::Use src(Range(), &(cast->gtOp.gtOp1), cast);
```
Comment (pgavlin) [inline src/jit/decomposelongs.cpp]: Similar nit as above.

**Comment by pgavlin** [comment]
LGTM aside from a couple nits re: comments.

File: src/jit/decomposelongs.cpp (hunk 12 above)
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: @pgavlin Rather than adding 2 new comments I reorganized the code so that the overflow case is handled first instead of being intermixed with the other 2 cases. Together with an already existing comment I think it should be clear enough.

Comment (pgavlin) [inline src/jit/decomposelongs.cpp]: I concur. Already gave it another look :)

**Comment by BruceForstall** [comment]
@mikedn Nice! Thanks for the work!
